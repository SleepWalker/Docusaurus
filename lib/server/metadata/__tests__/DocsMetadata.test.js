/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const DocsMetadata = require('../DocsMetadata');

afterEach(() => {
  jest.resetAllMocks();
});

test('metadata for simple docs', () => {
  const docsDir = `${__dirname}/__fixtures__/simple-docs`;
  const meta = new DocsMetadata({
    docsDir,
    translatedDir: '',
    versionFallback: {
      docData() {
        return [];
      },
      sidebarData() {
        return {};
      },
    },
    sidebars: {
      root: {
        Test: ['doc1', 'doc2'],
      },
    },
    translation: {
      enabled: false,
      getFileLanguage() {
        return null;
      },
      enabledLanguages() {
        return [
          {
            enabled: true,
            name: 'English',
            tag: 'en',
          },
        ];
      },
    },
    versioning: {
      enabled: false,
    },
    useEnglishUrl: false,
  });

  const result = meta.populate();

  expect(result).toMatchSnapshot();
});

test('wrong id format', () => {
  const docsDir = `${__dirname}/__fixtures__/invalid-docs`;
  const meta = new DocsMetadata({
    docsDir,
    translatedDir: '',
    versionFallback: {
      docData() {
        return [];
      },
      sidebarData() {
        return {};
      },
    },
    sidebars: {},
    translation: {
      enabled: false,
      getFileLanguage() {
        return null;
      },
      enabledLanguages() {
        return [
          {
            enabled: true,
            name: 'English',
            tag: 'en',
          },
        ];
      },
    },
    versioning: {
      enabled: false,
    },
    useEnglishUrl: false,
  });
  jest.spyOn(console, 'warn').mockImplementation(() => {});

  expect(meta.populate.bind(meta)).toThrowErrorMatchingSnapshot();
  expect(console.warn).toHaveBeenCalledWith(
    expect.stringMatching(/Header field "bad" in .* is not supported/)
  );
});

it('metadata for siteConfig.useEnglishUrl', () => {
  const docsDir = `${__dirname}/__fixtures__/simple-docs`;
  const meta = new DocsMetadata({
    docsDir,
    translatedDir: '',
    versionFallback: {
      docData() {
        return [];
      },
      sidebarData() {
        return {};
      },
    },
    sidebars: {
      root: {
        Test: ['doc1', 'doc2'],
      },
    },
    translation: {
      enabled: false,
      getFileLanguage() {
        return null;
      },
      enabledLanguages() {
        return [
          {
            enabled: true,
            name: 'English',
            tag: 'en',
          },
        ];
      },
    },
    versioning: {
      enabled: false,
    },
    useEnglishUrl: true,
  });

  const result = meta.populate();

  expect(result).toMatchSnapshot();
});

test('metadata with versioning', () => {
  const docsDir = `${__dirname}/__fixtures__/simple-docs`;
  const meta = new DocsMetadata({
    docsDir,
    translatedDir: '',
    versionFallback: {
      docData() {
        return [
          {
            id: 'en-version-1.3.1-translation',
            title: 'Translations & Localization',
            original_id: 'translation',
            source: 'version-1.2.1/guides-translation.md',
            permalink: 'docs/en/1.3.1/translation.html',
            localized_id: 'version-1.3.1-translation',
            language: 'en',
            version: '1.3.1',
          },
          {
            id: 'en-version-1.3.1-versioning',
            title: 'Versioning',
            original_id: 'versioning',
            source: 'version-1.2.0/guides-versioning.md',
            permalink: 'docs/en/1.3.1/versioning.html',
            localized_id: 'version-1.3.1-versioning',
            language: 'en',
            version: '1.3.1',
          },
          {
            id: 'en-version-1.3.1-no-sidebar',
            title: 'No Sidebar',
            original_id: 'no-sidebar',
            source: 'version-1.2.0/guides-no-sidebar.md',
            permalink: 'docs/en/1.3.1/no-sidebar.html',
            localized_id: 'version-1.3.1-no-sidebar',
            language: 'en',
            version: '1.3.1',
          },
        ];
      },
      sidebarData() {
        return {
          'version-1.3.1-docs': {
            Guides: [
              'version-1.3.1-navigation',
              'version-1.3.1-translation',
              'version-1.3.1-versioning',
            ],
          },
        };
      },
    },
    sidebars: {
      root: {
        Test: ['doc1', 'doc2'],
      },
    },
    translation: {
      enabled: false,
      getFileLanguage() {
        return null;
      },
      enabledLanguages() {
        return [
          {
            enabled: true,
            name: 'English',
            tag: 'en',
          },
        ];
      },
    },
    versioning: {
      enabled: true,
    },
    useEnglishUrl: false,
  });

  const result = meta.populate();

  expect(result).toMatchSnapshot();
});

test('metadata with translations', () => {
  const docsDir = `${__dirname}/__fixtures__/simple-docs`;
  const translatedDir = `${__dirname}/__fixtures__/translated-docs`;
  const meta = new DocsMetadata({
    docsDir,
    translatedDir,
    versionFallback: {
      docData() {
        return [];
      },
      sidebarData() {
        return {};
      },
    },
    sidebars: {
      root: {
        Test: ['doc1', 'doc2'],
      },
    },
    translation: {
      enabled: true,
      getFileLanguage(file) {
        return file.includes('/de/') ? 'de' : 'en';
      },
      enabledLanguages() {
        return [
          {
            enabled: true,
            name: 'English',
            tag: 'en',
          },
          {
            enabled: true,
            name: 'Deutsch',
            tag: 'de',
          },
        ];
      },
    },
    versioning: {
      enabled: false,
    },
  });

  const result = meta.populate();

  expect(result).toMatchSnapshot();
});

test('metadata with translations and versioning', () => {
  const docsDir = `${__dirname}/__fixtures__/simple-docs`;
  const translatedDir = `${__dirname}/__fixtures__/translated-docs`;
  const meta = new DocsMetadata({
    docsDir,
    translatedDir,
    versionFallback: {
      docData() {
        return [];
      },
      sidebarData() {
        return {};
      },
    },
    sidebars: {
      root: {
        Test: ['doc1', 'doc2'],
      },
    },
    translation: {
      enabled: true,
      getFileLanguage(file) {
        return file.includes('/de/') ? 'de' : 'en';
      },
      enabledLanguages() {
        return [
          {
            enabled: true,
            name: 'English',
            tag: 'en',
          },
          {
            enabled: true,
            name: 'Deutsch',
            tag: 'de',
          },
        ];
      },
    },
    versioning: {
      enabled: true,
    },
    useEnglishUrl: false,
  });

  const result = meta.populate();

  expect(result).toMatchSnapshot();
});
