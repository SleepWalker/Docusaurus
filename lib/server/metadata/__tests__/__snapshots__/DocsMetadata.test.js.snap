// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metadata for simple docs 1`] = `
Object {
  "doc1": Object {
    "category": "Test",
    "id": "doc1",
    "language": "en",
    "localized_id": "doc1",
    "next": "doc2",
    "next_id": "doc2",
    "next_title": "doc2",
    "permalink": "docs/doc1.html",
    "previous_title": undefined,
    "sidebar": "root",
    "source": "doc1.md",
    "title": "Document 1",
  },
  "doc2": Object {
    "category": "Test",
    "id": "doc2",
    "language": "en",
    "localized_id": "doc2",
    "next_title": undefined,
    "permalink": "docs/doc2.html",
    "previous": "doc1",
    "previous_id": "doc1",
    "previous_title": "Document 1",
    "sidebar": "root",
    "source": "doc2.markdown",
    "title": "doc2",
  },
  "doc3": Object {
    "id": "doc3",
    "language": "en",
    "localized_id": "doc3",
    "next_title": undefined,
    "permalink": "docs/doc3.html",
    "previous_title": undefined,
    "source": "doc3.md",
    "title": "doc3",
  },
  "sub/doc1": Object {
    "id": "sub/doc1",
    "language": "en",
    "localized_id": "sub/doc1",
    "next_title": undefined,
    "permalink": "docs/sub/doc1.html",
    "previous_title": undefined,
    "source": "sub/doc1.md",
    "title": "Document 1",
  },
}
`;

exports[`metadata for siteConfig.useEnglishUrl 1`] = `
Object {
  "doc1": Object {
    "category": "Test",
    "id": "doc1",
    "language": "en",
    "localized_id": "doc1",
    "next": "doc2",
    "next_id": "doc2",
    "next_title": "doc2",
    "permalink": "docs/en/doc1.html",
    "previous_title": undefined,
    "sidebar": "root",
    "source": "doc1.md",
    "title": "Document 1",
  },
  "doc2": Object {
    "category": "Test",
    "id": "doc2",
    "language": "en",
    "localized_id": "doc2",
    "next_title": undefined,
    "permalink": "docs/en/doc2.html",
    "previous": "doc1",
    "previous_id": "doc1",
    "previous_title": "Document 1",
    "sidebar": "root",
    "source": "doc2.markdown",
    "title": "doc2",
  },
  "doc3": Object {
    "id": "doc3",
    "language": "en",
    "localized_id": "doc3",
    "next_title": undefined,
    "permalink": "docs/en/doc3.html",
    "previous_title": undefined,
    "source": "doc3.md",
    "title": "doc3",
  },
  "sub/doc1": Object {
    "id": "sub/doc1",
    "language": "en",
    "localized_id": "sub/doc1",
    "next_title": undefined,
    "permalink": "docs/en/sub/doc1.html",
    "previous_title": undefined,
    "source": "sub/doc1.md",
    "title": "Document 1",
  },
}
`;

exports[`metadata with translations 1`] = `
Object {
  "de-de/doc1": Object {
    "id": "de-de/doc1",
    "language": "de",
    "localized_id": "de/doc1",
    "next_title": undefined,
    "permalink": "docs/de/de/doc1.html",
    "previous_title": undefined,
    "source": "de/doc1.md",
    "title": "[DE] Document 1",
  },
  "de-de/doc2": Object {
    "id": "de-de/doc2",
    "language": "de",
    "localized_id": "de/doc2",
    "next_title": undefined,
    "permalink": "docs/de/de/doc2.html",
    "previous_title": undefined,
    "source": "de/doc2.markdown",
    "title": "[DE] Document 2",
  },
  "de-doc1": Object {
    "category": "Test",
    "id": "de-doc1",
    "language": "de",
    "localized_id": "doc1",
    "next": "de-doc2",
    "next_id": "doc2",
    "next_title": "doc2",
    "permalink": "docs/de/doc1.html",
    "previous_title": undefined,
    "sidebar": "root",
    "source": "doc1.md",
    "title": "Document 1",
  },
  "de-doc2": Object {
    "category": "Test",
    "id": "de-doc2",
    "language": "de",
    "localized_id": "doc2",
    "next_title": undefined,
    "permalink": "docs/de/doc2.html",
    "previous": "de-doc1",
    "previous_id": "doc1",
    "previous_title": "Document 1",
    "sidebar": "root",
    "source": "doc2.markdown",
    "title": "doc2",
  },
  "de-doc3": Object {
    "id": "de-doc3",
    "language": "de",
    "localized_id": "doc3",
    "next_title": undefined,
    "permalink": "docs/de/doc3.html",
    "previous_title": undefined,
    "source": "doc3.md",
    "title": "doc3",
  },
  "de-sub/doc1": Object {
    "id": "de-sub/doc1",
    "language": "de",
    "localized_id": "sub/doc1",
    "next_title": undefined,
    "permalink": "docs/de/sub/doc1.html",
    "previous_title": undefined,
    "source": "sub/doc1.md",
    "title": "Document 1",
  },
  "en-doc1": Object {
    "category": "Test",
    "id": "en-doc1",
    "language": "en",
    "localized_id": "doc1",
    "next": "en-doc2",
    "next_id": "doc2",
    "next_title": "doc2",
    "permalink": "docs/en/doc1.html",
    "previous_title": undefined,
    "sidebar": "root",
    "source": "doc1.md",
    "title": "Document 1",
  },
  "en-doc2": Object {
    "category": "Test",
    "id": "en-doc2",
    "language": "en",
    "localized_id": "doc2",
    "next_title": undefined,
    "permalink": "docs/en/doc2.html",
    "previous": "en-doc1",
    "previous_id": "doc1",
    "previous_title": "Document 1",
    "sidebar": "root",
    "source": "doc2.markdown",
    "title": "doc2",
  },
  "en-doc3": Object {
    "id": "en-doc3",
    "language": "en",
    "localized_id": "doc3",
    "next_title": undefined,
    "permalink": "docs/en/doc3.html",
    "previous_title": undefined,
    "source": "doc3.md",
    "title": "doc3",
  },
  "en-sub/doc1": Object {
    "id": "en-sub/doc1",
    "language": "en",
    "localized_id": "sub/doc1",
    "next_title": undefined,
    "permalink": "docs/en/sub/doc1.html",
    "previous_title": undefined,
    "source": "sub/doc1.md",
    "title": "Document 1",
  },
}
`;

exports[`metadata with translations and versioning 1`] = `
Object {
  "de-de/doc1": Object {
    "id": "de-de/doc1",
    "language": "de",
    "localized_id": "de/doc1",
    "next_title": undefined,
    "permalink": "docs/de/next/de/doc1.html",
    "previous_title": undefined,
    "source": "de/doc1.md",
    "title": "[DE] Document 1",
    "version": "next",
  },
  "de-de/doc2": Object {
    "id": "de-de/doc2",
    "language": "de",
    "localized_id": "de/doc2",
    "next_title": undefined,
    "permalink": "docs/de/next/de/doc2.html",
    "previous_title": undefined,
    "source": "de/doc2.markdown",
    "title": "[DE] Document 2",
    "version": "next",
  },
  "de-doc1": Object {
    "category": "Test",
    "id": "de-doc1",
    "language": "de",
    "localized_id": "doc1",
    "next": "de-doc2",
    "next_id": "doc2",
    "next_title": "doc2",
    "permalink": "docs/de/next/doc1.html",
    "previous_title": undefined,
    "sidebar": "root",
    "source": "doc1.md",
    "title": "Document 1",
    "version": "next",
  },
  "de-doc2": Object {
    "category": "Test",
    "id": "de-doc2",
    "language": "de",
    "localized_id": "doc2",
    "next_title": undefined,
    "permalink": "docs/de/next/doc2.html",
    "previous": "de-doc1",
    "previous_id": "doc1",
    "previous_title": "Document 1",
    "sidebar": "root",
    "source": "doc2.markdown",
    "title": "doc2",
    "version": "next",
  },
  "de-doc3": Object {
    "id": "de-doc3",
    "language": "de",
    "localized_id": "doc3",
    "next_title": undefined,
    "permalink": "docs/de/next/doc3.html",
    "previous_title": undefined,
    "source": "doc3.md",
    "title": "doc3",
    "version": "next",
  },
  "de-sub/doc1": Object {
    "id": "de-sub/doc1",
    "language": "de",
    "localized_id": "sub/doc1",
    "next_title": undefined,
    "permalink": "docs/de/next/sub/doc1.html",
    "previous_title": undefined,
    "source": "sub/doc1.md",
    "title": "Document 1",
    "version": "next",
  },
  "en-doc1": Object {
    "category": "Test",
    "id": "en-doc1",
    "language": "en",
    "localized_id": "doc1",
    "next": "en-doc2",
    "next_id": "doc2",
    "next_title": "doc2",
    "permalink": "docs/en/next/doc1.html",
    "previous_title": undefined,
    "sidebar": "root",
    "source": "doc1.md",
    "title": "Document 1",
    "version": "next",
  },
  "en-doc2": Object {
    "category": "Test",
    "id": "en-doc2",
    "language": "en",
    "localized_id": "doc2",
    "next_title": undefined,
    "permalink": "docs/en/next/doc2.html",
    "previous": "en-doc1",
    "previous_id": "doc1",
    "previous_title": "Document 1",
    "sidebar": "root",
    "source": "doc2.markdown",
    "title": "doc2",
    "version": "next",
  },
  "en-doc3": Object {
    "id": "en-doc3",
    "language": "en",
    "localized_id": "doc3",
    "next_title": undefined,
    "permalink": "docs/en/next/doc3.html",
    "previous_title": undefined,
    "source": "doc3.md",
    "title": "doc3",
    "version": "next",
  },
  "en-sub/doc1": Object {
    "id": "en-sub/doc1",
    "language": "en",
    "localized_id": "sub/doc1",
    "next_title": undefined,
    "permalink": "docs/en/next/sub/doc1.html",
    "previous_title": undefined,
    "source": "sub/doc1.md",
    "title": "Document 1",
    "version": "next",
  },
}
`;

exports[`metadata with versioning 1`] = `
Object {
  "doc1": Object {
    "category": "Test",
    "id": "doc1",
    "language": "en",
    "localized_id": "doc1",
    "next": "doc2",
    "next_id": "doc2",
    "next_title": "doc2",
    "permalink": "docs/next/doc1.html",
    "previous_title": undefined,
    "sidebar": "root",
    "source": "doc1.md",
    "title": "Document 1",
    "version": "next",
  },
  "doc2": Object {
    "category": "Test",
    "id": "doc2",
    "language": "en",
    "localized_id": "doc2",
    "next_title": undefined,
    "permalink": "docs/next/doc2.html",
    "previous": "doc1",
    "previous_id": "doc1",
    "previous_title": "Document 1",
    "sidebar": "root",
    "source": "doc2.markdown",
    "title": "doc2",
    "version": "next",
  },
  "doc3": Object {
    "id": "doc3",
    "language": "en",
    "localized_id": "doc3",
    "next_title": undefined,
    "permalink": "docs/next/doc3.html",
    "previous_title": undefined,
    "source": "doc3.md",
    "title": "doc3",
    "version": "next",
  },
  "en-version-1.3.1-no-sidebar": Object {
    "id": "en-version-1.3.1-no-sidebar",
    "language": "en",
    "localized_id": "version-1.3.1-no-sidebar",
    "next_title": undefined,
    "original_id": "no-sidebar",
    "permalink": "docs/en/1.3.1/no-sidebar.html",
    "previous_title": undefined,
    "source": "version-1.2.0/guides-no-sidebar.md",
    "title": "No Sidebar",
    "version": "1.3.1",
  },
  "en-version-1.3.1-translation": Object {
    "category": "Guides",
    "id": "en-version-1.3.1-translation",
    "language": "en",
    "localized_id": "version-1.3.1-translation",
    "next": "version-1.3.1-versioning",
    "next_id": "versioning",
    "next_title": undefined,
    "original_id": "translation",
    "permalink": "docs/en/1.3.1/translation.html",
    "previous": "version-1.3.1-navigation",
    "previous_id": "navigation",
    "previous_title": undefined,
    "sidebar": "version-1.3.1-docs",
    "source": "version-1.2.1/guides-translation.md",
    "title": "Translations & Localization",
    "version": "1.3.1",
  },
  "en-version-1.3.1-versioning": Object {
    "category": "Guides",
    "id": "en-version-1.3.1-versioning",
    "language": "en",
    "localized_id": "version-1.3.1-versioning",
    "next_title": undefined,
    "original_id": "versioning",
    "permalink": "docs/en/1.3.1/versioning.html",
    "previous": "version-1.3.1-translation",
    "previous_id": "translation",
    "previous_title": undefined,
    "sidebar": "version-1.3.1-docs",
    "source": "version-1.2.0/guides-versioning.md",
    "title": "Versioning",
    "version": "1.3.1",
  },
  "sub/doc1": Object {
    "id": "sub/doc1",
    "language": "en",
    "localized_id": "sub/doc1",
    "next_title": undefined,
    "permalink": "docs/next/sub/doc1.html",
    "previous_title": undefined,
    "source": "sub/doc1.md",
    "title": "Document 1",
    "version": "next",
  },
}
`;

exports[`wrong id format 1`] = `"Document id cannot include \\"/\\"."`;
